File: 0576-out-of-boundary-paths/0576-out-of-boundary-paths.cs
public class Solution {
    int M;    
    int N;
    long[,,] Memo;
    const long MOD = 1000000007;
    
    public int FindPaths(int m, int n, int maxMove, int startRow, int startColumn) {
        M = m;
        N = n;
        Memo = new long[m, n, maxMove];
        for (int i = 0; i < m; ++i)
        {
            for (int j = 0; j < n; ++j)
            {
                for (int k = 0; k < maxMove; ++k)
                {
                    Memo[i, j, k] = -1;
                }
            }
        }
        
        
        return (int)GetPathsRecursively(startRow, startColumn, maxMove);
    }
    
    public long GetPathsRecursively(int r, int c, int stepLeft)
    {
        if (r < 0 || M == r || c < 0 || N == c)
            return 1;
        if (stepLeft == 0)
            return 0;
        
        --stepLeft;
        if (Memo[r, c, stepLeft] != -1)
            return Memo[r, c, stepLeft];
        long result = GetPathsRecursively(r+1, c, stepLeft) +   
                    GetPathsRecursively(r-1, c, stepLeft) +
                    GetPathsRecursively(r, c+1, stepLeft) +
                    GetPathsRecursively(r, c-1, stepLeft);
        Memo[r, c, stepLeft] = result % MOD;
        return Memo[r, c, stepLeft];
    }
}

File: 0576-out-of-boundary-paths/0576-out-of-boundary-paths.js
/**
 * @param {number} m
 * @param {number} n
 * @param {number} maxMove
 * @param {number} startRow
 * @param {number} startColumn
 * @return {number}
 */
var findPaths = function(m, n, maxMove, startRow, startColumn) {
    let MOD = 1000000007;
    if (maxMove === 0)
        return 0
    
    let memo = new Array(m)
    for(let i=0; i<m; ++i) {
        memo[i] = new Array(n)
        for (let j = 0; j < n; ++j)
            memo[i][j] = new Array(maxMove)
    }
    
    const recursive = (x, y, leftMoves) => {
        if (leftMoves >= 0) {
            if (x < 0 || x === m || y < 0 || y === n) {
                return 1
            }
            else if (memo[x][y][leftMoves] === undefined){
                let nextMove = leftMoves-1
                memo[x][y][leftMoves] = 
                    recursive(x+1, y, nextMove) +
                    recursive(x, y+1, nextMove) +
                    recursive(x-1, y, nextMove) +
                    recursive(x, y-1, nextMove)
            }
            return memo[x][y][leftMoves] % MOD
        }
        else {
            return 0
        }
    }
    return recursive(startRow, startColumn, maxMove)
};